boot/boot_serial/src/boot_serial.c

```diff
@@ -186,17 +187,34 @@ bs_list(char *buf, int len)
             if (flash_area_open(area_id, &fap)) {
                 continue;
             }
+#ifdef MCUBOOT_IMAGE_ACCESS_HOOKS
```

Comment:
Rather than make this conditional, can we just always have hooks, even if there is hook that just accesses the data directly?

Reply:
I think it takes more operations, depending on number of parameters the function takes.
Maybe inlined stubs could be optimize by compiler.

I kinda agree on that no-ifdefs policy as this annoys me in C a lot; the solution is to either use the `IS_ENABLED` type of macro
```
if (IS_ENABLED(MCUBOOT_IMAGE_ACCESS_HOOKS) && boot_read_image_header_hook(image_index, slot, &hdr) == BOOT_HOOK_REGULAR) {
...
}
```
like in Zephyr, but this also requires code to be specifically written as you need to have all definitions used within such `if` available till compiler is able to optimize them out.
Another option would be to define macros, when `MCUBOOT_IMAGE_ACCESS_HOOKS` is not defined, that pretend to be hook functions and return specific literal, but this has to be carefully done, for example:
```
#ifndef MCUBOOT_IMAGE_ACCESS_HOOKS
#define boot_read_image_header_hook(just, dump, it) SPECIFIC_LITERAL
#endif

int rc = boot_read_image_header_hook(image_index, slot, &hdr);
if (rc == NOT_THAT_SPECIFIC_LITERAL) {
  ... // should be optimized out by compiler when it figures out that rc is set to literal
}
```
and also requires all definitions used within the block to be defined during compilation.

Reply:
What about construction with macro like this.
``` C
#ifdef MCUBOOT_IMAGE_ACCESS_HOOKS
#define HOOK_CALL(f, ret, ret_defult, ...) \
    do { \
            ret = f(__VA_ARGS__); \
    } while (0)
#else
#define HOOK_CALL(f, ret, ret_defult, ...) \
    do { \
            ret = ret_default; \
    } while (0)
#endif
```

```suggestion
        HOOK_CALL(boot_read_image_header_hook, rc, BOOT_HOOK_REGULAR, BOOT_CURR_IMG(state), i, boot_img_hdr(state, i));
        if (rc == BOOT_HOOK_REGULAR)
        {
            flash_area_read(fap, 0, &hdr, sizeof(hdr));
        }
```

` if (rc == BOOT_HOOK_REGULAR)` can be embedded into HOOK_CALL macro if you are thinking that it is better.

Reply:
Introduced macro for call the hook which become no-op in case hooks are not enabled.